### PR TITLE
fix: let raw_pathway_df handle present headers but no data

### LINKED ISSUE #233: MEO Empty File Parse Output Error
In MEO’s parse_output, when the algorithm returns an empty raw pathway - represented by a file containing only the header `Source Type Target Oriented Weight` - the raw_pathway_df function in util.py should detect this using pd.errors.EmptyDataError (as described [here](https://pandas.pydata.org/docs/reference/api/pandas.errors.EmptyDataError.html)) and return an empty DataFrame with the SPRAS standard output columns. However, this behavior is not currently happening. When I check if df.empty outside of the raw_pathway_df function, the df is correctly identified as empty.

### PR DESCRIPTION
**Pull request by tristan-f-r**
~~(@ntalluri what exactly did you do for #233? I think this addresses a separate issue, but I'm not entirely sure.)~~

Closes #233.

### CODE REVIEW

File: spras/meo.py [hunk 1]
```diff
@@ -202,7 +202,10 @@ def parse_output(raw_pathway_file, standardized_pathway_file):
         """
         # Columns Source Type Target Oriented Weight
         df = raw_pathway_df(raw_pathway_file, sep='\t', header=0)
```
[the remark this answers is not in the document]
Comment (ntalluri) [reply spras/meo.py:204]: And the empty file is the raw-pathway with only a MEO specific header.

File: test/MEO/test_meo.py
[the remark this answers is not in the document]
Comment (ntalluri) [reply test/MEO/test_meo.py:1]: it has to do with the output being empty though, not the inputs given to meo as being empty. That's a separate issue all together.

[the remark this answers is not in the document]
Comment (ntalluri) [reply test/MEO/test_meo.py:1]: Please add this to the parse-output test suite. It inherently tests raw_pathway_df.

File: test/parse-outputs/expected/meo-present-headers-expected.txt
Comment (ntalluri) [inline test/parse-outputs/expected/meo-present-headers-expected.txt:1]: This file is equivalent to the file _"test/parse-outputs/expected/empty-pathway-expected.txt"_ which is the file format we expect every algorithm in SPRAS to use when the parsed output is empty. Please switch the test case to use the _empty-pathway-expected.txt_ file and remove this expected file.

File: test/parse-outputs/input/meo-present-headers.txt
Comment (ntalluri) [inline test/parse-outputs/input/meo-present-headers.txt:1]: Rename file to _"meo-empty-raw-pathway.txt"_

File: test/parse-outputs/test_parse_outputs.py
Comment (ntalluri) [inline test/parse-outputs/test_parse_outputs.py:1]: Currently we are testing a completely empty file. It would be useful to double check what all the empty files are for each of the algorithms to make sure that the empty file is being represented correctly and adding that to the input folder and updating the test cases in test_parse_outputs.py

OI2 has very weird output formats and will take a lot of testing to make sure we get all the correct empty cases. This issue has some of the information on the headers https://github.com/Reed-CompBio/spras/issues/133, but not all the information on what the empty files are.

File: test/parse-outputs/test_parse_outputs.py [hunk 2]
```diff
@@ -21,42 +23,49 @@ def setup_class(cls):
...
-            test_file = DUPLICATE_EDGES_DIR + f"{algo}-raw-pathway.txt"
-            out_file = OUTDIR + f"{algo}-duplicate-pathway.txt"
+            test_file = DUPLICATE_EDGES_DIR / f"{algo}-raw-pathway.txt"
+            out_file = OUTDIR / f"{algo}-duplicate-pathway.txt"
 
             runner.parse_output(algo, test_file, out_file)
-            assert filecmp.cmp(OUTDIR + f"{algo}-duplicate-pathway.txt", EXPDIR + f"{algo}-pathway-expected.txt", shallow=False)
+            assert filecmp.cmp(OUTDIR / f"{algo}-duplicate-pathway.txt", EXPDIR / f"{algo}-pathway-expected.txt", shallow=False)
+
+    def test_meo_present_headers(self):
```
Comment (ntalluri) [inline test/parse-outputs/test_parse_outputs.py:66]: Instead update the test_empty_file test case with MEO + other algorithms empty file formats.

File: spras/util.py [hunk 3]
```diff
@@ -72,6 +72,8 @@ def raw_pathway_df(raw_pathway_file: str, sep: str = '\t', header: int = None) -
     """
     try:
         df = pd.read_csv(raw_pathway_file, sep=sep, header=header)
+        if df.empty:
+            raise pd.errors.EmptyDataError(ValueError("Columns are present but there is no data to parse."))
```
[the remark this answers is not in the document]
Comment (ntalluri) [reply spras/util.py:76]: This "error" is likely to occur frequently, so it would be better to clarify the reason in a comment with specifics rather than printing out this message. 

In this case, it’s not truly a ValueError, the algorithm simply returned an empty output, which is often expected behavior. We don’t need to display a message here, since empty outputs from algorithms are valid and not necessarily indicative of a problem with the outputs of the algorithm. 
- this also will be shown that empty files are occurring in the summary.txt file if a user toggles it to true.

File: test/parse-outputs/test_parse_outputs.py [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
+import pytest
+
 from spras import runner
 
-INDIR = "test/parse-outputs/input/"
-OUTDIR = "test/parse-outputs/output/"
-EXPDIR = "test/parse-outputs/expected/"
-OI2_EDGE_CASES_INDIR = 'test/parse-outputs/input/omicsintegrator-edge-cases/'
-DUPLICATE_EDGES_DIR = 'test/parse-outputs/input/duplicate-edges/'
+INDIR = Path("test", "parse-outputs", "input")
```
Comment (ntalluri) [inline test/parse-outputs/test_parse_outputs.py:8]: I saw a lot of updates to the filepaths in many of the PRs, was there a reason for this?

File: test/parse-outputs/test_parse_outputs.py
Comment (ntalluri) [reply test/parse-outputs/test_parse_outputs.py:1]: The contributing guide should also be updated to ask a user to add in the empty file of the specific algorithm.

File: test/parse-outputs/expected/meo-present-headers-expected.txt
Comment (tristan-f-r) [reply test/parse-outputs/expected/meo-present-headers-expected.txt:1]: At the time of writing this PR I had a decently misguided understanding of the `parse-outputs` test folder, which has luckily changed since then but I completely forgot to update the rest of this PR to address that. I'll do that now and add the proper test cases to cover the range of errors :+1:

File: spras/util.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply spras/util.py]: This does do the same thing as leaving a comment, since the error message is never printed out (it is immediately caught). I wanted to do it this way to avoid duplicating the empty dataframe call later, but I've changed it to duplicate that line of code :+1:

File: test/parse-outputs/test_parse_outputs.py (hunk 4 above)
Comment (ntalluri) [reply test/parse-outputs/test_parse_outputs.py:8]: Is there a reason for doing this?

**Comment by tristan-f-r** [comment]
(currently trying to generate an empty DOMINO test file)

**Comment by tristan-f-r** [comment]
@ntalluri good catch! DOMINO also would have had a similar error.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -205,7 +205,8 @@ Modify generate inputs:
 Modify parse outputs:
 1. Obtain the raw-pathway output (e.g. from the run function in your wrapper by running the Snakemake workflow) and save it to `test/parse-outputs/input`. Name it as `{algorithm_name}-raw-pathway.txt`.
 2. Obtain the expected universal output from the workflow, manually confirm it is correct, and save it to `test/parse-outputs/expected` directory. Name it as `{algorithm_name}-pathway-expected.txt`.
-3. Add the new algorithm's name to the algorithms list in `test/parse-outputs/test_parse_outputs.py`.
+3. Add an `{algorithm-name}-empty-raw-pathway.txt` file inside `test/parse-outputs/input/empty` containing an output file with no useful graph output, with any decorating information (e.g. headers)
```
Comment (ntalluri) [inline CONTRIBUTING.md:208]: Step 3 is confusing, what is no useful graph output mean? This step should be obtain the file that is associated with an empty subnetwork/output returned by the algortihm.

File: test/parse-outputs/input/empty/omicsintegrator2-empty-raw-pathway.txt
Comment (ntalluri) [inline test/parse-outputs/input/empty/omicsintegrator2-empty-raw-pathway.txt:1]: There are couple options that occur for the empty raw pathway for OI2.

File: spras/domino.py [hunk 6]
```diff
@@ -192,8 +192,10 @@ def parse_output(raw_pathway_file, standardized_pathway_file):
                     # columns that indicate edges
                     # Dropping the other rows eliminates the node information
                     module_df = pd.DataFrame(entries)
-                    module_df = module_df.loc[:, ['source', 'target']].dropna()
-
+                    try:
+                        module_df = module_df.loc[:, ['source', 'target']].dropna()
+                    except KeyError:
+                        module_df = pd.DataFrame()
```
Comment (ntalluri) [inline spras/domino.py:198]: There are no headers for this? What is edges_df concatting on?

Comment (tristan-f-r) [reply spras/domino.py:198]: I've solely removed any edges from DOMINO with "source" and "target" - all other data is preserved. 

Note that `pd.Concat` allows for empty dataframes:

```
>>> pandas.concat([pandas.DataFrame(), pandas.DataFrame()], axis=0)
Empty DataFrame
Columns: []
Index: []
```

File: test/parse-outputs/input/empty/omicsintegrator2-empty-raw-pathway-1.txt
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply test/parse-outputs/input/empty/omicsintegrator2-empty-raw-pathway-1.txt:1]: I have no clue. My OI2 runs have not told me what it is.

File: test/parse-outputs/input/empty/omicsintegrator2-empty-raw-pathway-2.txt
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply test/parse-outputs/input/empty/omicsintegrator2-empty-raw-pathway-2.txt:1]: I can't find any more, though I don't really have much to test with.

File: test/parse-outputs/test_parse_outputs.py [hunk 7]
```diff
@@ -21,42 +24,52 @@ def setup_class(cls):
...
-            test_file = INDIR + f"{algo}-raw-pathway.txt"
-            out_file = OUTDIR + f"{algo}-pathway.txt"
+            test_file = INDIR / f"{algo}-raw-pathway.txt"
+            out_file = OUTDIR / f"{algo}-pathway.txt"
 
             runner.parse_output(algo, test_file, out_file)
-            assert filecmp.cmp(OUTDIR + f"{algo}-pathway.txt", EXPDIR + f"{algo}-pathway-expected.txt", shallow=False)
+            assert filecmp.cmp(OUTDIR / f"{algo}-pathway.txt", EXPDIR / f"{algo}-pathway-expected.txt", shallow=False)
 
     def test_empty_file(self):
```
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply test/parse-outputs/test_parse_outputs.py:37]: (I am going to assume that this is definitionally equal to the last comment.)

**Comment by tristan-f-r** [comment]
If more OI2 test cases appear, I will add them. I can not find any more, and the fact that we don't have any issues reporting other OI2 test cases means that most likely these edge cases are caught.

[I can also add some more if anyone has datasets that would bring about these types of errors, but I do not have these.]

**Comment by ntalluri** [comment]
https://github.com/Reed-CompBio/spras/issues/133#issuecomment-2321590234

**Comment by ntalluri** [comment]
s

**Comment by ntalluri** [comment]
> If more OI2 test cases appear, I will add them. I can not find any more, and the fact that we don't have any issues reporting other OI2 test cases means that most likely these edge cases are caught.
> 
> [I can also add some more if anyone has datasets that would bring about these types of errors, but I do not have these.]

Have you tried running a large test sweep on egfr?

**Comment by tristan-f-r** [comment]
[I forgot to put in my commit for the protein1-protein2 dataset]

**Comment by ntalluri** [comment]
I think that is the only one i have run into as well

File: test/parse-outputs/input/empty/omicsintegrator2-empty-raw-pathway-1.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+some one line error (TODO: can we reproduce the exact error? I can't.)
```
Comment (ntalluri) [inline test/parse-outputs/input/empty/omicsintegrator2-empty-raw-pathway-1.txt:1]: Is the empty file just a file with a single empty line?

Comment (tristan-f-r) [reply test/parse-outputs/input/empty/omicsintegrator2-empty-raw-pathway-1.txt:1]: Perhaps!

File: test/parse-outputs/input/empty/omicsintegrator2-empty-raw-pathway-1.txt
Comment (ntalluri) [inline test/parse-outputs/input/empty/omicsintegrator2-empty-raw-pathway-1.txt:1]: should this be an empty file with on line? I think that is what the code in Oi2 parse output is referring to. I also haven't ran into this error in a while...

File: test/parse-outputs/expected/meo-present-headers-expected.txt
Comment (ntalluri) [inline test/parse-outputs/expected/meo-present-headers-expected.txt:1]: This is the expected header for every output. Why is there a separate one for MEO? Is this also for the empty case?

File: spras/domino.py (hunk 6 above)
Comment (ntalluri) [inline spras/domino.py:198]: Just to confirm, this is creating an empty df because there are no source/target columns that indicate edges? And ere the potential empty cases are no source/target keys and/or source/target columns with no content?

**Comment by tristan-f-r** [comment]
I also ran into this error during benchmarking - having this patch would be nice :+1:

**Comment by tristan-f-r** [comment]
https://github.com/tristan-f-r/spras/blob/e2d184b2559618e609bd8e796b99d4abac6ad756/test/parse-outputs/input/empty/bowtiebuilder-empty-raw-pathway.txt

File: test/parse-outputs/test_parse_outputs.py
Comment (ntalluri) [inline test/parse-outputs/test_parse_outputs.py:1]: since you added in the new test case for testing all empty files, I think we can remove this test case

```
 def test_empty_file(self):
        for algo, params in algorithms.items():
            test_file = INDIR / f"empty-raw-pathway.txt"
            out_file = OUTDIR / f"{algo}-empty-pathway.txt"
            runner.parse_output(algo, test_file, out_file, params)
            assert filecmp.cmp(OUTDIR / f"{algo}-empty-pathway.txt", EXPDIR / f"empty-pathway-expected.txt", shallow=False)
```

File: test/parse-outputs/test_parse_outputs.py [hunk 9]
```diff
@@ -89,3 +92,12 @@ def test_duplicate_edges(self):
 
             runner.parse_output(algo, test_file, out_file, params)
             assert filecmp.cmp(OUTDIR / f"{algo}-duplicate-pathway.txt", EXPDIR / f"{algo}-pathway-expected.txt", shallow=False)
+
+    def test_empty_raw_pathway(self):
+        for algo, params in algorithms.items():
+            files = glob(str(INDIR / "empty" / f"{algo}-empty-raw-pathway*"))
+            assert len(files) != 0, "can't test on no empty raw pathways!"
+            for test_file in glob(str(INDIR / "empty" / f"{algo}-empty-raw-pathway*")):
+                out_file = OUTDIR / f"{algo}-empty-raw-pathway-output.txt"
```
Comment (ntalluri) [inline test/parse-outputs/test_parse_outputs.py:101]: This output file should be `f"{algo}-empty-pathway-output.txt"` since it is converting the raw-pathway into the standarized pathway formart

File: test/parse-outputs/test_parse_outputs.py [hunk 10]
```diff
@@ -89,3 +92,12 @@ def test_duplicate_edges(self):
 
             runner.parse_output(algo, test_file, out_file, params)
             assert filecmp.cmp(OUTDIR / f"{algo}-duplicate-pathway.txt", EXPDIR / f"{algo}-pathway-expected.txt", shallow=False)
+
+    def test_empty_raw_pathway(self):
+        for algo, params in algorithms.items():
+            files = glob(str(INDIR / "empty" / f"{algo}-empty-raw-pathway*"))
+            assert len(files) != 0, "can't test on no empty raw pathways!"
```
Comment (ntalluri) [inline test/parse-outputs/test_parse_outputs.py:99]: is this incase all the files get deleted?

**Review by ntalluri — CHANGES_REQUESTED** [review]
I am going to test this locally but I have a few changes/questions for the test_parse_outputs.py file.

File: test/parse-outputs/test_parse_outputs.py [hunk 11]
```diff
@@ -87,7 +87,7 @@ def test_duplicate_edges(self):
 
     def test_empty_raw_pathway(self):
         for algo, params in algorithms.items():
-            files = glob(str(INDIR / "empty" / f"{algo}-empty-raw-pathway*"))
+            files = glob(str(INDIR / "empty" / f"{algo}-empty-pathway*"))
```
Comment (ntalluri) [inline test/parse-outputs/test_parse_outputs.py:90]: This was correct before, for the out_file it should be changed to f"{algo}-empty-pathway-output.txt"

Comment (ntalluri) [reply test/parse-outputs/test_parse_outputs.py:90]: The files for the input are the empty-raw-pathway files
